keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,282 @@
+
+### Non-Goals
+
+   * Providing alternate models of IP allocation schemes for pod CIDR.
+   * Enhancing current allocators to handle disjoint ranges.
+   * Enhancing current allocators to add additional ranges after cluster creation.
+   * Changing current assumptions around having a single pod CIDR per node.
+
+## Proposal
+
+As stated above, the goal is to re-implement the functionality called out in the summary, but in a 
+way that does not depend on a pod cluster CIDR. The essence of the proposal is that for the 
+first two cases in iptables implementation and first case in ipvs, we can replace the `-s proxier.clusterCIDR` with some notion of node local pod traffic.
+
+The core logic in these cases is “how to determine” cluster originated traffic from non-cluster originated ones. The proposal is that tracking pod traffic generated from within the node is sufficient to determine cluster originated traffic.
+For the first two use cases in iptables and first use case in ipvs, we provide alternatives to using the proxier.clusterCIDR in one of the following ways to determine cluster originated traffic
+
+   1. `-s node.podCIDR` (where node podCIDR is used for allocating pod IPs within the node)
+   2. `--in-interface prefix+` (where all pod interfaces start with same prefix)
+   3. `-m physdev --physdev-is-in` (for kubenet if we don’t want to depend on node podCIDR)
+
+Note the above are equivalent definitions, when considering only pod traffic originating from within the node.
+
+For the last use case, in iptables and ipvs, the proposal is to drop the reference to the cluster CIDR.
+
+The reasoning behind why this works are as follows.
+
+### iptables - masquerade off cluster traffic to services by node IP
+
+The [rule here currently](
+  https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L887-L894
+) looks as follows
+
+```go
+// This masquerades off-cluster traffic to a service VIP.  The idea
+// is that you can establish a static route for your Service range,
+// routing to any node, and that node will bridge into the Service
+// for you.  Since that might bounce off-node, we masquerade here.
+// If/when we support "Local" policy for VIPs, we should update this.
+writeLine(proxier.natRules, append(args, "! -s", proxier.clusterCIDR, "-j", string(KubeMarkMasqChain))...)
+```
+
+The logic is that if the source IP is not part of the cluster CIDR range,
+then it must have originated from outside the cluster. Hence we add a rule to masquerade by
+the node IP so that we can send traffic to any pod within the cluster.
+
+One key insight when thinking about this data path though is the fact that the iptable rules run
+at _every_ node boundary. So when a pod sends a traffic to a service IP, it gets translated to
+one of the node IPs _before_ it leaves the node at the node boundary. So it's highly unlikely to 
+receive traffic at a node, whose destination is the service cluster IP, that is initiated by pods
```

Comment:
In addition, to dovetail with the Ingress V2 work that @bowei is driving, is it reasonable to assume that providing external access to service should be thought of in that context and we mostly look at "internal routing" as far as kube-proxy is concerned?

Reply:
Thanks for this use case. I was thinking about it a bit. I think the important rule is that the return path has to hit the node where the DNAT from svc to pod IP happened (as that has the conntrack enry) or otherwise the reverse-DNAT will not happen on the return packet. Without the reverse-DNAT, the originating pod will drop the return packet as it will not match the 5-tuple. So In the case you described where we have pod (A) in node A sending traffic to node B, which does the DNAT to pod (C) in some node C, we have (A) -> B -> (C). Now the return path from (C) has to hit B or the reverse-DNAT will not happen. For that to be true, the traffic from B -> (C), in addition to the DNAT also needs to SNAT by node B IP, right? So in this case, the proposed change in even more correct as node B will correctly SNAT all traffic that did not originate from node B. Is my understanding right?

Reply:
Thanks for this use case. I was thinking about it a bit. I think the important rule is that the return path has to hit the node where the DNAT from svc to pod IP happened (as that has the conntrack enry) or otherwise the reverse-DNAT will not happen on the return packet. Without the reverse-DNAT, the originating pod will drop the return packet as it will not match the 5-tuple. So In the case you described where we have pod (A) in node A sending traffic to node B, which does the DNAT to pod (C) in some node C, we have (A) -> B -> (C). Now the return path from (C) has to hit B or the reverse-DNAT will not happen. For that to be true, the traffic from B -> (C), in addition to the DNAT also needs to SNAT by node B IP, right? So in this case, the proposed change in even more correct as node B will correctly SNAT all traffic that did not originate from node B. Is my understanding right?

Reply:
given the above understanding, should this be resolved? or am I missing something?